### DISCUSSION TITLE
Output every word in a separate line

### DISCUSSION THREAD

**Discussion by Joelson-Forte** [Purfview/whisper-standalone-win#107] [Ideas]
Hi,
Is it possible to add an option to generate word by word separately with their respective duration equal to Buzz "Word Level Timings" option?

![2023-11-13_06h51_53](https://github.com/Purfview/whisper-standalone-win/assets/131564527/1508e41e-c866-4c9a-8621-3bf377c532ae)

I found this option to be very useful, as if you use Subtitle Edit Tools > Merge Short Lines option on the generated file, you can create a default subtitle by choosing the maximum characters per line. The most incredible thing is that the synchronization is practically perfect, obeying all the pauses. I think this happens because Subtitle Edit takes into account the duration of each word to create the default subtitle.

The only problem I found is that some words have very short durations. Some words even have a duration of 0.000 and to fix this I had to make a custom script that adjusts all times with a duration of less than 100ms to 100ms.

See my screen recording to understand better. I've attached the files I used in the test in case you need them.
https://drive.google.com/file/d/14CtJvPXTxREUtyGrtx47wPf0mXdfaSxV/view?usp=sharing

I think this option in Whisper Faster would be useful because Buzz is very limited and does not offer the option of command lines for the user to specify the variables they want.

Thanks!

**Comment by Purfview** [comment]
This option is already in Faster-Whisper, and it's enabled by default. Just that it doesn't output every single word in new lines.

**Reply by Joelson-Forte** [reply]
Exactly. If a file is generated with all the words on new lines, it will be possible to create subtitles in Subtitle Edit according to the user's needs. Many people complain about the long lines generated by Whisper and when they try to correct this in Subtitle Edit, the synchronization is not very good. But using the file with all the words on new lines this problem does not occur, and if it does occur, it is minimal, easy to correct.

If you do some tests with some files you will notice how the subtitles are well synchronized using the file that contains all the words on new lines. As I said before, the only problem I found are words with very short durations (less than 100ms or 0.000ms) that need to be corrected.

**Reply by Purfview** [reply]
I'll look what can be done.

>  ...correct this in Subtitle Edit, the synchronization is not very good.

Do they/you use "post-processing" in SE for that? I never use "post-processing" as it includes bunch of various unneeded "fixes". I apply only exact fixes I need with "Fix common errors" & ect.

**Reply by Joelson-Forte** [reply]
To be honest I don't understand much about Subtitle Edit software either. In fact, I only discovered it was possible to create subtitles using the file Word Level Timing of the BUS in Subtitle Edit, because I was testing the Subtitle Edit options.

I was looking for a way to solve the problem of very long lines created in Whisper and at the same time find a way to define the max number of characters and lines in the subtitle while maintaining synchronization with the audio file.

I was testing Whisper --max_line_width MAX_LINE_WIDTH and --max_line_count MAX_LINE_COUNT commands and saw that the result it wasn't good, impossible to use. However, if we knew how this Subtitle Edit software mechanism with the Bus Word Level Timing file works, it might be possible to do this directly in Whisper-Faster without the need to create a file with all the words on new lines. It would be enough to enable the --max_line_width MAX_LINE_WIDTH and --max_line_count MAX_LINE_COUNT command lines to create a file using this mechanism.

**Reply by Purfview** [reply]
Dunno if it's worth to bother, check subs on your "tigers" file: https://we.tl/t-tyxPp4t6iN

Sub 1 - made from "all words separate". [your file]
Sub 2 - made from "words not separate".

**Reply by Joelson-Forte** [reply]
I understand. I would think so too seeing these results.
In any case, I would like to thank you for taking the time to analyze this request.

**Reply by Purfview** [reply]
It would bring a bit more accuracy only in a middle of split lines.

I've looked at the existing code and looks like it should be easy hack to support such feature, maybe in few hours I'll compile the new release.

**Reply by Joelson-Forte** [reply]
Looking forward to seeing how it turns out. 
Hoping that everything goes well and Whisper Faster becomes even better than it already is.

**Comment by Purfview** [comment]
Implemented in `r160.5` as `--one_word` switch.
Btw, removes spaces too.

**Reply by Joelson-Forte** [reply]
It turned out very good. See in my screen recording how I use the --one_word file in editing.
https://drive.google.com/file/d/1SjbHzS27ztVB_k4DYUTRInqFvFYja5cn/view?usp=sharing

Thank you for making this child's day happier.

**Reply by Purfview** [reply]
I see you found a way how to makes the proper highlighted karaoke subs.
So you don't need that borked way anymore -> [Max Line Width and Max Line Count Commands](https://github.com/Purfview/whisper-standalone-win/discussions/87), no?

Good that I was lazy to enable those commands for you, despite that it's easy. 😄

**Reply by Joelson-Forte** [reply]
I never got good synchronization with these commands and that's why I gave up on them. Fortunately I found this way that uses the words on a separate line, which so far works very well.

The only problem with this procedure, as I said before, is words with very short durations (less than 34ms). Words with a duration of less than 34ms cannot be imported into most video editors, which have a minimum standard of 1 frame (34ms) to import srt files.

For example, in the video you saw, lines 26, 118 and 128 will not be imported as they last less than 34ms.
![2023-11-14_18h44_56](https://github.com/Purfview/whisper-standalone-win/assets/131564527/490cc340-039c-445e-a3f4-6cde902031b0)

I have a script made by a friend that adjusts timings lasting less than 100ms in the srt file. It finds all words less than 100ms and leaves it at 100ms. I just find it annoying having to put <u> before each word and </u> after each word. 😄😄😄😄😄

I'm sending you the script. There is also a screen recording showing how to use it. It was designed to work directly in the Video Editor. But you will certainly understand how it works much better than I did and maybe this can help others who also need it.

https://drive.google.com/file/d/16c3kbI7fDTIagz_6nsgi3CpcRdc31k4C/view?usp=sharing

**Reply by Purfview** [reply]
>I have a script made by a friend that adjusts timings lasting less than 100ms in the srt file. 

Your friend instead of writing whole 200 lines script could just edit few lines in existing whisper's [func ](https://github.com/openai/whisper/blob/1cea4357687b676b293cb5473e1ade25f5b1cef7/whisper/utils.py#L187). 😄

> It finds all words less than 100ms and leaves it at 100ms.

Did it made 100ms by overlapping adjacent subs?

>  I just find it annoying having to put before each word and after each word. 

Put what?

EDIT:
Strange that Subtitle Edit can't fix those lines by extending duration backwards, did you asked about that in SE repo?
I think, I could make some simple hack to not output <50ms subs and not to overlap anything.

**Comment by Joelson-Forte** [comment]
> Did it made 100ms by overlapping adjacent subs?

I'm not sure, but I think what's added in one line is made up for in subsequent lines. As it is only 100ms it is practically imperceptible.
> 
> > I just find it annoying having to put before each word and after each word.
> Put what?

These special characters. I noticed that neither Google Translate nor Git Hub make these characters appear. I show this in the screen recording I sent.
![Untitled-1](https://github.com/Purfview/whisper-standalone-win/assets/131564527/d1be7d43-5d7d-436e-8926-d68066aed67c)

> EDIT: Strange that Subtitle Edit can't fix those lines by extending duration backwards, did you asked about that in SE repo? I think, I could make some simple hack to not output <50ms subs and not to overlap anything.

I made a request but received no response 
https://github.com/SubtitleEdit/subtitleedit/discussions/7561

If you can fix this it will be a great help and one less script in the lives of those who edit subtitles. 50ms is enough 😄

**Reply by Purfview** [reply]
> nor Git Hub make these characters appear.

Easy -> `</u>`.
Press quote this post to see how to do it.

> If you can fix this it will be a great help and one less script in the lives of those who edit subtitles. 50ms is enough.

Piece of cake.

**Reply by Joelson-Forte** [reply]
Ahhhh I now undestand. Is the option Code <>. This is all new to me.😄

I want to take advantage and ask a question: Do you know of any command that creates a highlight box like the one I made in my screen recording in the file generated by the --highlight_words option?

I want to replace the commands `<u>` and `</u>` (default commands that generate an underline in each word) with a command that creates a highlight box in each word.

For example, currently I can replace `<u>` and `</u>` with the commands {\1c&H3E8D05&} and {\r} that generate the green highlight color in each word. But I don't know how to create a highlight box. In fact, I don't even know if it's possible.

**Reply by Purfview** [reply]
> Do you know of any command that creates a highlight box..

I don't know such. That in VEGAS? Maybe it's just in VEGAS and it's not any subtitle feature. 

> For example, currently I can replace <u> and </u> with the commands {\1c&H3E8D05&} and {\r} ...

Those "commands" are the tags,
"{\1c&H3E8D05&} and {\r}" <- this doesn't look like for srt files.

**Reply by Joelson-Forte** [reply]
See this video to undestand better
https://drive.google.com/file/d/1kXtprO5b9ALmRzRVknPWWgBYJ_XI7o2c/view?usp=sharing

I replaced the original tags `<u> and </u>` from the srt file generated with the "--highlight_words true" option, with the tags `{\1c&H53C600&} and {\r}` By doing this the underline of each word was replaced by the green highlight in each word.

I was looking for a tag that create a highlight box for each word in srt file. I looked for this information a lot, but I couldn't find it. I just thought maybe you knew.

**Reply by Purfview** [reply]
If a one player supports non-standard tags doesn't mean that other support them too.

**Comment by Joelson-Forte** [comment]
> If a one player supports non-standard tags doesn't mean that other support them too.

To solve this I convert the .srt to .ass format or burn the .ass directly into the video with ffmpeg.

**Comment by Joelson-Forte** [comment]
> EDIT: Strange that Subtitle Edit can't fix those lines by extending duration backwards, did you asked about that in SE repo? I think, I could make some simple hack to not output <50ms subs and not to overlap anything.

Did you manage to do this hack to adjust the very short duration times? I tested version 160.6 and there are still words with a duration of less than 34ms in --one_word srt and vtt file. I only mentioned this subject again so you don't forget.😄

**Comment by Purfview** [comment]
Check this test build with upgraded  `--one_word` : https://we.tl/t-VoxK3Y9jyt
There shouldn't be lines shorter than 50 ms nor overlaps.

If you see an issue then post .srt and .json files.

**Reply by justdoc1** [reply]
I'm getting an error when I just enter '--one_word' on the command line. What else do I need to enter?

**Reply by Purfview** [reply]
Read help. `--help`

**Comment by Joelson-Forte** [comment]
I did several tests. It has improved a lot, but words lasting less than 50ms still occur.

Here is one of the tests I did. There is a .txt file informing which lines the very short durations occur on. It also has the results before modification to compare the results.
https://drive.google.com/file/d/1fZX5Fl7xKEmins6zs5AvEKeTnkdbRo2G/view?usp=sharing

I noticed that this trial version only produces srt files. I tried outputting to other formats but it only produces srt files. That's why I didn't send the files in .json format.

The .json files that are attached are from the models I used.

**Reply by Purfview** [reply]
> I noticed that this trial version only produces srt files. I tried outputting to other formats but it only produces srt files.

I've no idea how this is possible, post your command.

**Reply by Joelson-Forte** [reply]
Example 1
`C:\Users\Joelson\Desktop\whisper-faster.exe.lnk "C:\Users\Joelson\Desktop\my video or audio" --model medium --one_word --output_format vtt`

Example 2
`C:\Users\Joelson\Desktop\whisper-faster.exe.lnk "C:\Users\Joelson\Desktop\my video or audio" --model medium --one_word --output_format srt`

Example 3
`C:\Users\Joelson\Desktop\whisper-faster.exe.lnk "C:\Users\Joelson\Desktop\my video or audio" --model medium --one_word --output_format all`

Example 4
`C:\Users\Joelson\Desktop\whisper-faster.exe.lnk "C:\Users\Joelson\Desktop\my video or audio" --model medium --output_format all`

Note: Whisper-Faster_r160.6 works fine here

**Reply by Joelson-Forte** [reply]
Restart the PC and reinstall everything again. Now it appears to be working correctly. I'm going to do new tests now. It was definitely something I did wrong.

**Reply by Purfview** [reply]
Where did you look for subtitles? By your examples they should be at .exe file. [not at shortcut]
And at the end of operation there is written path to subtitles. I need only json, no need for anything else.

**Reply by Joelson-Forte** [reply]
I'm redoing the test now. I'm using this command line and everything worked fine.
`"C:\Users\Joelson\Desktop\whisper-faster.exe.lnk" "C:\Users\Joelson\Desktop\my vieo or audio" --model medium --one_word --output_format all`
When finished, I will send you the .json files

**Reply by Joelson-Forte** [reply]
Here are the srt and json files of the test
https://drive.google.com/file/d/1HVeSbq9-A240ZKXJYb9Gh9W8Yg1c8pkW/view?usp=sharing

Lines with duration less than 50ms
197
462
889
1297
1699
1779

**Reply by Purfview** [reply]
"No access" in that link.

**Reply by Joelson-Forte** [reply]
Try now
https://drive.google.com/file/d/1HVeSbq9-A240ZKXJYb9Gh9W8Yg1c8pkW/view?usp=sharing

**Reply by Purfview** [reply]
New test build, hopefully no more <50 ms nor overlaps ->  https://we.tl/t-13ojURVn2x

**Reply by Joelson-Forte** [reply]
Now it's good. I did three tests, one of which was a 1 hour and 20 minute video. I didn't find any lines shorter than 50ms.

The srt and json files are at this link if you need them.
https://drive.google.com/file/d/1mzsutOY0AV9bBXXIxYG42x4fN7Up7a3X/view?usp=sharing

Your work is truly incredible. Thank you very much!

**Reply by Purfview** [reply]
Do we need to save the old method without time corrections as an option?

**Reply by Joelson-Forte** [reply]
I think it's a good idea to save a Backup of the old method. Until we have time and absolute certainty that the new method is reliable. This can only be measured over time and with users feedback. You agree?

**Reply by Joelson-Forte** [reply]
I'm now editing the 1 hour and 20 minute video and I can assure you that the subtitles generated by your application are now far superior to the subtitles generated by Adobe Premiere. There's almost nothing to fix.

The synchronization with the audio file is impeccable and the accuracy in transcribing the words is impressive.

Since most of my video editing work consists of subtitling, I think I got an early end of year gift.😄

**Reply by Purfview** [reply]
Nice.
You could make a video example of  how to assemble these one_word subs and make highlighted subs.

**Reply by Joelson-Forte** [reply]
Are you talking about those highlighted subtitles I create in VEGAS?

**Reply by Purfview** [reply]
That too.

**Reply by Joelson-Forte** [reply]
OK here is the video. Please understand that this effect does not exist natively in VEGAS. I had to create the custom scripts and presets to be able to create this effect in VEGAS.
https://drive.google.com/file/d/165fA-l_Fz-vccvAzKfB260-vnYTUwt3h/view?usp=sharing

If you want, I can make a compilation with all the material needed to be used in VEGAS. You just need to use the installer and everything will automatically be placed in its right place. But you will need to have VEGAS installed on your system.

**Reply by Purfview** [reply]
Oh, maybe don't bother then.
I personally don't need it, thought that some users could be interested.

**Reply by Joelson-Forte** [reply]
OK. If you need anything and I can help, just get in touch.

Thank you again

**Comment by jianming0** [comment]
> OK here is the video. Please understand that this effect does not exist natively in VEGAS. I had to create the custom scripts and presets to be able to create this effect in VEGAS. https://drive.google.com/file/d/165fA-l_Fz-vccvAzKfB260-vnYTUwt3h/view?usp=sharing
> 
> If you want, I can make a compilation with all the material needed to be used in VEGAS. You just need to use the installer and everything will automatically be placed in its right place. But you will need to have VEGAS installed on your system.

Hi Jedit, saw your video on the word highlights which I thought was pretty darn cool. Am trying to figure out something similar for Davinci Resolve if you'd be kind enough to shed some light on the scripting logic behind what your script is actually doing to the 2 subtitles layers?

**Reply by Joelson-Forte** [reply]
Hi jianming0,

1.On the top track is the caption with all the individual words.

2.On the bottom track is the standard subtitle (Single Line) with a maximum of 20 characters.

This value of 20 characters will be used for both the 16x9 (Horizontal) and the 9x16 Vertical format. But this value can be different as long as everything is in a single caption line.

3.You will notice that the text events on the top track correspond with the text events on the bottom track. The difference is that in the upper track the words are individual and in the lower track the words are joined together to form the sentence.

4.When selecting the text event from the bottom track, the script checks how many words there are in the selected text event and adjusts the positioning of the words in the text events on the top track to be in the same position as the respective words on the bottom track.

**Comment by matheusunderline** [comment]
Cool discussion, :sweat_smile: !

Do you have some plan to update the Linux version with the new features showed here, Purfview?

Thank you to develop the _whisper_standalone_win_ :heart: !

**Reply by Purfview** [reply]
> Do you have some plan to update the Linux version...

Why, no one is using it anyway... -> https://github.com/Purfview/whisper-standalone-win/discussions/109

Jk, yes, eventually I'll update it.

**Reply by matheusunderline** [reply]
Oh, I didn't know about that. I just changed to _GNU/Linux_ (_KDE Neon_ distribution) some days.

Anyway, I guess that I can visualize effort on something none take advantage. Maybe, I will try use the latest _Windows_ version with wine or something like that in _GNU/Linux_.

Thank you to fast reply! :heart:
